### PR TITLE
Feature/37 - Get engagements between dates endpoint

### PR DESCRIPTION
**Pull request by tjanowsk**
I didn't realise this while creating the branch, but this is actually #37 .

This is needed for the pie chart.

Note that I have changed the endpoint to truncate the engagements if they cross the query time, e.g. if there is an engagement between 1pm-3pm and the query specifies the start time at 2pm, the returning engagement will be 2pm-3pm.
This will allow to calculate the times more precisely.

### CODE REVIEW

File: backend/src/main/java/com/rota/api/StaffService.java [hunk 1]
```diff
@@ -32,12 +69,26 @@
    * @return A list of member's engagements
    */
   public List<EngagementDto> getStaffEngagementsBetween(int staffId, Instant start, Instant end) {
-    Instant startTime = start == null ? Instant.MIN : start;
-    Instant endTime = end == null ? Instant.MAX : end;
     return engagementRepository.findByStaffId(staffId).stream()
```
Comment (emab) [inline backend/src/main/java/com/rota/api/StaffService.java:72]: How do we think performance will be affected when someone has worked somewhere for a few years? Could be lots of engagements, will this cause issues? Just curious.

File: backend/src/main/java/com/rota/api/StaffService.java [hunk 2]
```diff
@@ -21,6 +24,40 @@
...
+
+    return (Engagement engagement) -> !engagement.getEnd().isBefore(startTime)
+        && !engagement.getStart().isAfter(endTime);
+  }
+
+  /**
+   * Truncates the engagement, e.g. if if starts before the start time the function will cut off the
+   * time difference and set the engagement start time to requested start time.
+   */
+  private Function<Engagement, Engagement> truncateEngagement(Instant start, Instant end) {
```
Comment (emab) [inline backend/src/main/java/com/rota/api/StaffService.java:46]: I like how you've taken these out of the lambda expressions, I didn't know you could do this in Java.

File: backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package com.rota.exceptions;
 
 public class StaffNotFoundException extends RuntimeException {
+  static final long serialVersionUID = 1L;
```
Comment (emab) [inline backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java:4]: What does this do?

File: backend/src/test/java/com/rota/api/EngagementTests.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.time.Instant;
+import java.util.List;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class EngagementTests {
```
Comment (emab) [inline backend/src/test/java/com/rota/api/EngagementTests.java:20]: Good effort adding tests, I need to add some for methods I've created but I've been a bit lazy!

We could add tests for no engagements being returned and some negative tests such as invalid dates, start date after end date etc.

File: backend/src/main/java/com/rota/api/StaffService.java [hunk 5]
```diff
@@ -32,12 +69,26 @@
...
-                && !engagement.getEnd().isAfter(endTime))
+        .filter(isEngagementBetween(start, end))
+        .map(truncateEngagement(start, end))
+        .map(EngagementDto::fromEngagement)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Returns all engagements between start and end dates (inclusive).
+   * Specifically engagements that start after the start date and end before the end date.
```
Comment (emab) [inline backend/src/main/java/com/rota/api/StaffService.java:81]: I think you changed this so that it gets engagements that overlap now as well?

File: backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java (hunk 3 above)
Comment (tjanowsk) [reply backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java:4]: I was getting a warning that a Serializable class needs a serialVersionUID, so this helps to suppress that.

Some explanation here
https://stackoverflow.com/questions/285793/what-is-a-serialversionuid-and-why-should-i-use-it#285809

File: backend/src/main/java/com/rota/api/StaffService.java (hunk 1 above)
Comment (tjanowsk) [reply backend/src/main/java/com/rota/api/StaffService.java:72]: Possibly, depends on company's retention policy for engagements.
Unfortunately, I can't think of another way of writing this at the moment.

File: backend/src/main/java/com/rota/api/StaffService.java (hunk 5 above)
Comment (tjanowsk) [reply backend/src/main/java/com/rota/api/StaffService.java:81]: Yes, forgot to update the comment

File: backend/src/main/java/com/rota/api/StaffService.java (hunk 2 above)
Comment (tjanowsk) [reply backend/src/main/java/com/rota/api/StaffService.java:46]: Everything in java is a class, `lambda`s are no exception.
Under the hood lambda expressions are classes from `java.util.function`.

Not sure if this is the most OO way of handling this, I may have been corrupted by functional programming.

File: backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java (hunk 3 above)
Comment (emab) [reply backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java:4]: Ah right didn't realise `RuntimeException` implemented `Serializable`!

Out of interest where did you see the warning?

File: backend/src/main/java/com/rota/api/StaffService.java (hunk 2 above)
Comment (emab) [reply backend/src/main/java/com/rota/api/StaffService.java:46]: I did some reading on it and some places were suggesting that a lambda shouldn't be too long otherwise it gets too complicated, so taking this logic out of there makes sense. If we needed to reuse `isEngagementBetween` in a different scenario a boolean might have been better, but we don't so I think this way is good!

File: backend/src/main/java/com/rota/api/StaffService.java (hunk 1 above)
Comment (emab) [reply backend/src/main/java/com/rota/api/StaffService.java:72]: Yeah neither can I.... I don't know enough about databases to know the performance hit.

The only think I can think is that the generated SQL for this will end up always getting all engagement records from the database and _then_ performing a filter. Is there a way to construct an SQL statement that would avoid this? It's probably not part of MVP to get this bit right though.

File: backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java (hunk 3 above)
Comment (tjanowsk) [reply backend/src/main/java/com/rota/exceptions/StaffNotFoundException.java:4]: Neovim with coc.nvim plugin complains.
Given that this plugin is basically vscode's intellisense, I think vscode will complain too.

File: backend/src/main/java/com/rota/api/StaffService.java (hunk 1 above)
Comment (tjanowsk) [reply backend/src/main/java/com/rota/api/StaffService.java:72]: It should be possible to construct an sql statement which returns engagements between the dates.
I'll look into this, not sure if I'll make it part of this ticket though.
